src/upnp.cpp

```diff
@@ -965,8 +965,31 @@ void upnp::on_upnp_xml(error_code const& e
 		return;
 	}
 
+	// determine IGD version
+	std::string igd_str(p.get_body().begin, p.get_body().end);
+	size_t igd_ver_idx = igd_str.find("urn:schemas-upnp-org:device:InternetGatewayDevice:");
+	if (igd_ver_idx == std::string::npos)
+	{
+		char msg[500];
+		snprintf(msg, sizeof(msg), "missing IGD version information from %s"
+			, d.url.c_str());
+		//log(msg, l);
+		return;
+	}
+	int igd_ver = igd_str[igd_ver_idx + 50] - '0';
+	if (igd_ver < 1 || igd_ver > 2)
+	{
+		char msg[500];
+		snprintf(msg, sizeof(msg), "invaild IGD version from %s, falling back to version 1"
+			, d.url.c_str());
+		log(msg, l);
+		igd_ver = 1;
+	}
+	d.igd_version = igd_ver;
+
 	parse_state s;
-	s.reset("urn:schemas-upnp-org:service:WANIPConnection:1");
+	static std::string wanip_service = std::string("urn:schemas-upnp-org:service:WANIPConnection:") + to_string(d.igd_version).elems;
```

Comment:
> I think it would make more sense find_control_url to recognize the string prefix and parse out the version itself. It may even be a good opportunity to merge looking for WANIPConnection and PPPConnection as well.

I don't see any easy way to achieve this, unless a rewrite of `find_control_url`, what do you suggest to do that?

> Even if those two paths aren't merged, doesn't PPPConnection have to support version 2 as well?

No, from [InternetGatewayDevice:2  Device Template Version 1.01](http://upnp.org/specs/gw/UPnP-gw-InternetGatewayDevice-v2-Device.pdf),
```
1.2. Focus and Goals for DCP version 2.0 
Note: Due to lack of interest,  WANPPPConnection service was NOT updated to reflect the changes made to WANIPConnection. 
```
so `WANPPPConnection` only have version 1.